File: android/app/src/test/java/ai/oriveo/community/core/model/FolderColorTest.kt
package ai.oriveo.community.core.model

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Test

class FolderColorTest {

    @Test
    fun `all 10 colors exist`() {
        assertEquals(10, FolderColor.entries.size)
    }

    @Test
    fun `fromTag returns correct color`() {
        assertEquals(FolderColor.BLUE, FolderColor.fromTag("blue"))
        assertEquals(FolderColor.PINK, FolderColor.fromTag("pink"))
        assertEquals(FolderColor.GRAY, FolderColor.fromTag("gray"))
    }

    @Test
    fun `fromTag returns BLUE for null`() {
        assertEquals(FolderColor.BLUE, FolderColor.fromTag(null))
    }

    @Test
    fun `fromTag returns BLUE for unknown tag`() {
        assertEquals(FolderColor.BLUE, FolderColor.fromTag("neon"))
    }

    @Test
    fun `nextColor returns blue for empty list`() {
        assertEquals(FolderColor.BLUE, FolderColor.nextColor(emptyList()))
    }

    @Test
    fun `nextColor returns purple after blue`() {
        val folders = listOf(
            Folder(id = "1", name = "A", sortOrder = 1000, colorTag = "blue"),
        )
        assertEquals(FolderColor.PURPLE, FolderColor.nextColor(folders))
    }

    @Test
    fun `nextColor wraps around after gray`() {
        val folders = listOf(
            Folder(id = "1", name = "A", sortOrder = 1000, colorTag = "gray"),
        )
        assertEquals(FolderColor.BLUE, FolderColor.nextColor(folders))
    }

    @Test
    fun `nextColor uses last folder by sortOrder`() {
        val folders = listOf(
            Folder(id = "1", name = "A", sortOrder = 2000, colorTag = "red"),
            Folder(id = "2", name = "B", sortOrder = 1000, colorTag = "blue"),
            Folder(id = "3", name = "C", sortOrder = 3000, colorTag = "green"),
        )
        assertEquals(FolderColor.TEAL, FolderColor.nextColor(folders))
    }

    @Test
    fun `sequential 10 folders cycle full palette`() {
        val expected = FolderColor.entries.map { it.tag }
        val folders = mutableListOf<Folder>()
        for (i in 0 until 10) {
            val color = FolderColor.nextColor(folders)
            folders.add(
                Folder(
                    id = "f$i",
                    name = "Folder $i",
                    sortOrder = (i + 1) * 1000,
                    colorTag = color.tag,
                ),
            )
        }
        assertEquals(expected, folders.map { it.colorTag })
    }

    @Test
    fun `toColor returns non-null`() {
        for (fc in FolderColor.entries) {
            assertNotNull(fc.toColor())
        }
    }

    @Test
    fun `gradientBrush returns non-null`() {
        for (fc in FolderColor.entries) {
            assertNotNull(fc.gradientBrush())
        }
    }
}

File: web/apps/app/lib/core/__tests__/folder-color.test.ts
import { describe, it, expect } from 'vitest';
import {
  FOLDER_COLORS,
  FOLDER_COLOR_ORDER,
  getFolderColorPair,
  getNextFolderColor,
} from '@oriveo/shared';
import type { Folder } from '@oriveo/shared';

function makeFolder(overrides: Partial<Folder> = {}): Folder {
  return {
    id: crypto.randomUUID(),
    name: 'Test',
    sortOrder: 1000,
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString(),
    ...overrides,
  };
}

describe('folder-color', () => {
  describe('FOLDER_COLORS', () => {
    it('has 10 colors', () => {
      expect(Object.keys(FOLDER_COLORS)).toHaveLength(10);
    });

    it('each color has main and dark hex values', () => {
      for (const [, { main, dark }] of Object.entries(FOLDER_COLORS)) {
        expect(main).toMatch(/^#[0-9A-F]{6}$/i);
        expect(dark).toMatch(/^#[0-9A-F]{6}$/i);
      }
    });
  });

  describe('FOLDER_COLOR_ORDER', () => {
    it('has 10 entries matching FOLDER_COLORS keys', () => {
      expect(FOLDER_COLOR_ORDER).toHaveLength(10);
      for (const tag of FOLDER_COLOR_ORDER) {
        expect(FOLDER_COLORS[tag]).toBeDefined();
      }
    });

    it('starts with blue', () => {
      expect(FOLDER_COLOR_ORDER[0]).toBe('blue');
    });
  });

  describe('getFolderColorPair', () => {
    it('returns blue pair for undefined', () => {
      const [main, dark] = getFolderColorPair(undefined);
      expect(main).toBe(FOLDER_COLORS['blue'].main);
      expect(dark).toBe(FOLDER_COLORS['blue'].dark);
    });

    it('returns blue pair for unrecognized tag', () => {
      const [main] = getFolderColorPair('nonexistent');
      expect(main).toBe(FOLDER_COLORS['blue'].main);
    });

    it('returns correct pair for valid tag', () => {
      const [main, dark] = getFolderColorPair('pink');
      expect(main).toBe(FOLDER_COLORS['pink'].main);
      expect(dark).toBe(FOLDER_COLORS['pink'].dark);
    });
  });

  describe('getNextFolderColor', () => {
    it('returns blue for empty array', () => {
      expect(getNextFolderColor([])).toBe('blue');
    });

    it('returns purple after blue', () => {
      const folders = [makeFolder({ sortOrder: 1000, colorTag: 'blue' })];
      expect(getNextFolderColor(folders)).toBe('purple');
    });

    it('wraps around after gray', () => {
      const folders = [makeFolder({ sortOrder: 1000, colorTag: 'gray' })];
      expect(getNextFolderColor(folders)).toBe('blue');
    });

    it('uses the last folder by sortOrder', () => {
      const folders = [
        makeFolder({ sortOrder: 2000, colorTag: 'red' }),
        makeFolder({ sortOrder: 1000, colorTag: 'blue' }),
        makeFolder({ sortOrder: 3000, colorTag: 'green' }),
      ];
      expect(getNextFolderColor(folders)).toBe('teal'); // next after green
    });

    it('returns blue when last folder has no colorTag', () => {
      const folders = [makeFolder({ sortOrder: 1000 })];
      expect(getNextFolderColor(folders)).toBe('blue');
    });

    it('returns blue when last folder has unknown colorTag', () => {
      const folders = [makeFolder({ sortOrder: 1000, colorTag: 'neon' })];
      expect(getNextFolderColor(folders)).toBe('blue');
    });
  });
});
